File: GestureLock/src/main/java/com/atguigu/gesturelock/widget/GestureDrawline.java
package com.atguigu.gesturelock.widget;

import android.content.Context;
import android.graphics.Bitmap;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.graphics.PorterDuff;
import android.os.Handler;
import android.util.Pair;
import android.view.MotionEvent;
import android.view.View;

import com.atguigu.gesturelock.common.AppUtil;
import com.atguigu.gesturelock.common.Constants;
import com.atguigu.gesturelock.entity.GesturePoint;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * Created by Administrator on 2016/10/31.
 */

public class GestureDrawline extends View {
    private int mov_x;// 声明起点坐标
    private int mov_y;
    private Paint paint;// 声明画笔
    private Canvas canvas;// 画布
    private Bitmap bitmap;// 位图
    private List<GesturePoint> list;// 装有各个view坐标的集合
    private List<Pair<GesturePoint, GesturePoint>> lineList;// 记录画过的线
    private Map<String, GesturePoint> autoCheckPointMap;// 自动选中的情况点
    private boolean isDrawEnable = true; // 是否允许绘制

    /**
     * 屏幕的宽度和高度
     */
    private int[] screenDispaly;

    /**
     * 手指当前在哪个Point内
     */
    private GesturePoint currentPoint;
    /**
     * 用户绘图的回调
     */
    private GestureCallBack callBack;

    /**
     * 用户当前绘制的图形密码
     */
    private StringBuilder passWordSb;

    /**
     * 是否为校验
     */
    private boolean isVerify;

    /**
     * 用户传入的passWord
     */
    private String passWord;

    public GestureDrawline(Context context, List<GesturePoint> list, boolean isVerify,
                           String passWord, GestureCallBack callBack) {
        super(context);
        screenDispaly = AppUtil.getScreenDispaly(context);
        paint = new Paint(Paint.DITHER_FLAG);// 创建一个画笔
        bitmap = Bitmap.createBitmap(screenDispaly[0], screenDispaly[0], Bitmap.Config.ARGB_8888); // 设置位图的宽高
        canvas = new Canvas();
        canvas.setBitmap(bitmap);
        paint.setStyle(Paint.Style.STROKE);// 设置非填充
        paint.setStrokeWidth(10);// 笔宽5像素
        paint.setColor(Color.rgb(245, 142, 33));// 设置默认连线颜色
        paint.setAntiAlias(true);// 不显示锯齿

        this.list = list;
        this.lineList = new ArrayList<Pair<GesturePoint, GesturePoint>>();

        initAutoCheckPointMap();
        this.callBack = callBack;

        // 初始化密码缓存
        this.isVerify = isVerify;
        this.passWordSb = new StringBuilder();
        this.passWord = passWord;
    }

    private void initAutoCheckPointMap() {
        autoCheckPointMap = new HashMap<String,GesturePoint>();
        autoCheckPointMap.put("1,3", getGesturePointByNum(2));
        autoCheckPointMap.put("1,7", getGesturePointByNum(4));
        autoCheckPointMap.put("1,9", getGesturePointByNum(5));
        autoCheckPointMap.put("2,8", getGesturePointByNum(5));
        autoCheckPointMap.put("3,7", getGesturePointByNum(5));
        autoCheckPointMap.put("3,9", getGesturePointByNum(6));
        autoCheckPointMap.put("4,6", getGesturePointByNum(5));
        autoCheckPointMap.put("7,9", getGesturePointByNum(8));
    }

    private GesturePoint getGesturePointByNum(int num) {
        for (GesturePoint point : list) {
            if (point.getNum() == num) {
                return point;
            }
        }
        return null;
    }

    // 画位图
    @Override
    protected void onDraw(Canvas canvas) {
        // super.onDraw(canvas);
        canvas.drawBitmap(bitmap, 0, 0, null);
    }

    // 触摸事件
    @Override
    public boolean onTouchEvent(MotionEvent event) {
        if (isDrawEnable == false) {
            // 当期不允许绘制
            return true;
        }
        paint.setColor(Color.rgb(245, 142, 33));// 设置默认连线颜色
        switch (event.getAction()) {
            case MotionEvent.ACTION_DOWN:
                mov_x = (int) event.getX();
                mov_y = (int) event.getY();
                // 判断当前点击的位置是处于哪个点之内
                currentPoint = getPointAt(mov_x, mov_y);
                if (currentPoint != null) {
                    currentPoint.setPointState(Constants.POINT_STATE_SELECTED);
                    passWordSb.append(currentPoint.getNum());
                }
                // canvas.drawPoint(mov_x, mov_y, paint);// 画点
                invalidate();
                break;
            case MotionEvent.ACTION_MOVE:
                clearScreenAndDrawList();

                // 得到当前移动位置是处于哪个点内
                GesturePoint pointAt = getPointAt((int) event.getX(), (int) event.getY());
                // 代表当前用户手指处于点与点之前
                if (currentPoint == null && pointAt == null) {
                    return true;
                } else {// 代表用户的手指移动到了点上
                    if (currentPoint == null) {// 先判断当前的point是不是为null
                        // 如果为空，那么把手指移动到的点赋值给currentPoint
                        currentPoint = pointAt;
                        // 把currentPoint这个点设置选中为true;
                        currentPoint.setPointState(Constants.POINT_STATE_SELECTED);
                        passWordSb.append(currentPoint.getNum());
                    }
                }
                if (pointAt == null || currentPoint.equals(pointAt) || Constants.POINT_STATE_SELECTED == pointAt.getPointState()) {
                    // 点击移动区域不在圆的区域，或者当前点击的点与当前移动到的点的位置相同，或者当前点击的点处于选中状态
                    // 那么以当前的点中心为起点，以手指移动位置为终点画线
                    canvas.drawLine(currentPoint.getCenterX(), currentPoint.getCenterY(), event.getX(), event.getY(), paint);// 画线
                } else {
                    // 如果当前点击的点与当前移动到的点的位置不同
                    // 那么以前前点的中心为起点，以手移动到的点的位置画线
                    canvas.drawLine(currentPoint.getCenterX(), currentPoint.getCenterY(), pointAt.getCenterX(), pointAt.getCenterY(), paint);// 画线
                    pointAt.setPointState(Constants.POINT_STATE_SELECTED);

                    // 判断是否中间点需要选中
                    GesturePoint betweenPoint = getBetweenCheckPoint(currentPoint, pointAt);
                    if (betweenPoint != null && Constants.POINT_STATE_SELECTED != betweenPoint.getPointState()) {
                        // 存在中间点并且没有被选中
                        Pair<GesturePoint, GesturePoint> pair1 = new Pair<GesturePoint, GesturePoint>(currentPoint, betweenPoint);
                        lineList.add(pair1);
                        passWordSb.append(betweenPoint.getNum());
                        Pair<GesturePoint, GesturePoint> pair2 = new Pair<GesturePoint, GesturePoint>(betweenPoint, pointAt);
                        lineList.add(pair2);
                        passWordSb.append(pointAt.getNum());
                        // 设置中间点选中
                        betweenPoint.setPointState(Constants.POINT_STATE_SELECTED);
                        // 赋值当前的point;
                        currentPoint = pointAt;
                    } else {
                        Pair<GesturePoint, GesturePoint> pair = new Pair<GesturePoint, GesturePoint>(currentPoint, pointAt);
                        lineList.add(pair);
                        passWordSb.append(pointAt.getNum());
                        // 赋值当前的point;
                        currentPoint = pointAt;
                    }
                }
                invalidate();
                break;
            case MotionEvent.ACTION_UP:// 当手指抬起的时候
                if (isVerify) {
                    // 手势密码校验
                    // 清掉屏幕上所有的线，只画上集合里面保存的线
                    if (passWord.equals(passWordSb.toString())) {
                        // 代表用户绘制的密码手势与传入的密码相同
                        callBack.checkedSuccess();
                    } else {
                        // 用户绘制的密码与传入的密码不同。
                        callBack.checkedFail();
                    }
                } else {
                    callBack.onGestureCodeInput(passWordSb.toString());
                }
                break;
            default:
                break;
        }
        return true;
    }

    /**
     * 指定时间去清除绘制的状态
     * @param delayTime 延迟执行时间
     */
    public void clearDrawlineState(long delayTime) {
        if (delayTime > 0) {
            // 绘制红色提示路线
            isDrawEnable = false;
            drawErrorPathTip();
        }
        new Handler().postDelayed(new clearStateRunnable(), delayTime);
    }

    /**
     * 清除绘制状态的线程
     */
    final class clearStateRunnable implements Runnable {
        public void run() {
            // 重置passWordSb
            passWordSb = new StringBuilder();
            // 清空保存点的集合
            lineList.clear();
            // 重新绘制界面
            clearScreenAndDrawList();
            for (GesturePoint p : list) {
                p.setPointState(Constants.POINT_STATE_NORMAL);
            }
            invalidate();
            isDrawEnable = true;
        }
    }

    /**
     * 通过点的位置去集合里面查找这个点是包含在哪个Point里面的
     *
     * @param x
     * @param y
     * @return 如果没有找到，则返回null，代表用户当前移动的地方属于点与点之间
     */
    private GesturePoint getPointAt(int x, int y) {

        for (GesturePoint point : list) {
            // 先判断x
            int leftX = point.getLeftX();
            int rightX = point.getRightX();
            if (!(x >= leftX && x < rightX)) {
                // 如果为假，则跳到下一个对比
                continue;
            }

            int topY = point.getTopY();
            int bottomY = point.getBottomY();
            if (!(y >= topY && y < bottomY)) {
                // 如果为假，则跳到下一个对比
                continue;
            }

            // 如果执行到这，那么说明当前点击的点的位置在遍历到点的位置这个地方
            return point;
        }

        return null;
    }

    private GesturePoint getBetweenCheckPoint(GesturePoint pointStart, GesturePoint pointEnd) {
        int startNum = pointStart.getNum();
        int endNum = pointEnd.getNum();
        String key = null;
        if (startNum < endNum) {
            key = startNum + "," + endNum;
        } else {
            key = endNum + "," + startNum;
        }
        return autoCheckPointMap.get(key);
    }

    /**
     * 清掉屏幕上所有的线，然后画出集合里面的线
     */
    private void clearScreenAndDrawList() {
        canvas.drawColor(Color.TRANSPARENT, PorterDuff.Mode.CLEAR);
        for (Pair<GesturePoint, GesturePoint> pair : lineList) {
            canvas.drawLine(pair.first.getCenterX(), pair.first.getCenterY(),
                    pair.second.getCenterX(), pair.second.getCenterY(), paint);// 画线
        }
    }

    /**
     * 校验错误/两次绘制不一致提示
     */
    private void drawErrorPathTip() {
        canvas.drawColor(Color.TRANSPARENT, PorterDuff.Mode.CLEAR);
        paint.setColor(Color.rgb(154, 7, 21));// 设置默认线路颜色
        for (Pair<GesturePoint, GesturePoint> pair : lineList) {
            pair.first.setPointState(Constants.POINT_STATE_WRONG);
            pair.second.setPointState(Constants.POINT_STATE_WRONG);
            canvas.drawLine(pair.first.getCenterX(), pair.first.getCenterY(),
                    pair.second.getCenterX(), pair.second.getCenterY(), paint);// 画线
        }
        invalidate();
    }


    public interface GestureCallBack {

        /**
         * 用户设置/输入了手势密码
         */
        public abstract void onGestureCodeInput(String inputCode);

        /**
         * 代表用户绘制的密码与传入的密码相同
         */
        public abstract void checkedSuccess();

        /**
         * 代表用户绘制的密码与传入的密码不相同
         */
        public abstract void checkedFail();
    }
}

File: app/src/main/java/com/xpf/p2p/widget/GestureDrawline.kt
package com.xpf.p2p.widget

import android.content.Context
import android.graphics.Bitmap
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Paint
import android.graphics.PorterDuff
import android.os.Handler
import android.util.Pair
import android.view.MotionEvent
import android.view.View
import com.xpf.p2p.constants.Constants
import com.xpf.p2p.entity.GesturePoint
import com.xpf.p2p.utils.AppUtil

/**
 * Created by Administrator on 2016/10/31.
 */
class GestureDrawline(
    context: Context,
    private val list: List<GesturePoint>,
    private val isVerify: Boolean,
    private val passWord: String,
    private val callBack: GestureCallBack
) : View(context) {

    private var mov_x: Int = 0
    private var mov_y: Int = 0
    private val paint: Paint
    private val canvas: Canvas
    private val bitmap: Bitmap
    private val lineList: MutableList<Pair<GesturePoint, GesturePoint>>
    private val autoCheckPointMap: MutableMap<String, GesturePoint?>
    private var isDrawEnable = true
    private val screenDispaly: IntArray = AppUtil.getScreenDisplay(context)
    private var currentPoint: GesturePoint? = null
    private var passWordSb: StringBuilder

    init {
        paint = Paint(Paint.DITHER_FLAG)
        bitmap = Bitmap.createBitmap(screenDispaly[0], screenDispaly[0], Bitmap.Config.ARGB_8888)
        canvas = Canvas()
        canvas.setBitmap(bitmap)
        paint.style = Paint.Style.STROKE
        paint.strokeWidth = 10f
        paint.color = Color.rgb(245, 142, 33)
        paint.isAntiAlias = true

        lineList = ArrayList()
        autoCheckPointMap = HashMap()
        initAutoCheckPointMap()
        passWordSb = StringBuilder()
    }

    private fun initAutoCheckPointMap() {
        autoCheckPointMap["1,3"] = getGesturePointByNum(2)
        autoCheckPointMap["1,7"] = getGesturePointByNum(4)
        autoCheckPointMap["1,9"] = getGesturePointByNum(5)
        autoCheckPointMap["2,8"] = getGesturePointByNum(5)
        autoCheckPointMap["3,7"] = getGesturePointByNum(5)
        autoCheckPointMap["3,9"] = getGesturePointByNum(6)
        autoCheckPointMap["4,6"] = getGesturePointByNum(5)
        autoCheckPointMap["7,9"] = getGesturePointByNum(8)
    }

    private fun getGesturePointByNum(num: Int): GesturePoint? {
        for (point in list) {
            if (point.num == num) {
                return point
            }
        }
        return null
    }

    override fun onDraw(canvas: Canvas) {
        canvas.drawBitmap(bitmap, 0f, 0f, null)
    }

    override fun onTouchEvent(event: MotionEvent): Boolean {
        if (!isDrawEnable) {
            return true
        }
        paint.color = Color.rgb(245, 142, 33)
        when (event.action) {
            MotionEvent.ACTION_DOWN -> {
                mov_x = event.x.toInt()
                mov_y = event.y.toInt()
                currentPoint = getPointAt(mov_x, mov_y)
                if (currentPoint != null) {
                    currentPoint!!.pointState = Constants.POINT_STATE_SELECTED
                    passWordSb.append(currentPoint!!.num)
                }
                invalidate()
            }
            MotionEvent.ACTION_MOVE -> {
                clearScreenAndDrawList()
                val pointAt = getPointAt(event.x.toInt(), event.y.toInt())
                if (currentPoint == null && pointAt == null) {
                    return true
                } else {
                    if (currentPoint == null) {
                        currentPoint = pointAt
                        currentPoint!!.pointState = Constants.POINT_STATE_SELECTED
                        passWordSb.append(currentPoint!!.num)
                    }
                }
                if (pointAt == null || currentPoint == pointAt || Constants.POINT_STATE_SELECTED == pointAt.pointState) {
                    canvas.drawLine(
                        currentPoint!!.centerX.toFloat(), currentPoint!!.centerY.toFloat(),
                        event.x, event.y, paint
                    )
                } else {
                    canvas.drawLine(
                        currentPoint!!.centerX.toFloat(), currentPoint!!.centerY.toFloat(),
                        pointAt.centerX.toFloat(), pointAt.centerY.toFloat(), paint
                    )
                    pointAt.pointState = Constants.POINT_STATE_SELECTED

                    val betweenPoint = getBetweenCheckPoint(currentPoint!!, pointAt)
                    if (betweenPoint != null && Constants.POINT_STATE_SELECTED != betweenPoint.pointState) {
                        val pair1 = Pair(currentPoint!!, betweenPoint)
                        lineList.add(pair1)
                        passWordSb.append(betweenPoint.num)
                        val pair2 = Pair(betweenPoint, pointAt)
                        lineList.add(pair2)
                        passWordSb.append(pointAt.num)
                        betweenPoint.pointState = Constants.POINT_STATE_SELECTED
                        currentPoint = pointAt
                    } else {
                        val pair = Pair(currentPoint!!, pointAt)
                        lineList.add(pair)
                        passWordSb.append(pointAt.num)
                        currentPoint = pointAt
                    }
                }
                invalidate()
            }
            MotionEvent.ACTION_UP -> {
                if (isVerify) {
                    if (passWord == passWordSb.toString()) {
                        callBack.checkedSuccess()
                    } else {
                        callBack.checkedFail()
                    }
                } else {
                    callBack.onGestureCodeInput(passWordSb.toString())
                }
            }
        }
        return true
    }

    fun clearDrawlineState(delayTime: Long) {
        if (delayTime > 0) {
            isDrawEnable = false
            drawErrorPathTip()
        }
        Handler().postDelayed(ClearStateRunnable(), delayTime)
    }

    internal inner class ClearStateRunnable : Runnable {
        override fun run() {
            passWordSb = StringBuilder()
            lineList.clear()
            clearScreenAndDrawList()
            for (p in list) {
                p.pointState = Constants.POINT_STATE_NORMAL
            }
            invalidate()
            isDrawEnable = true
        }
    }

    private fun getPointAt(x: Int, y: Int): GesturePoint? {
        for (point in list) {
            val leftX = point.leftX
            val rightX = point.rightX
            if (!(x >= leftX && x < rightX)) {
                continue
            }
            val topY = point.topY
            val bottomY = point.bottomY
            if (!(y >= topY && y < bottomY)) {
                continue
            }
            return point
        }
        return null
    }

    private fun getBetweenCheckPoint(pointStart: GesturePoint, pointEnd: GesturePoint): GesturePoint? {
        val startNum = pointStart.num
        val endNum = pointEnd.num
        val key = if (startNum < endNum) "$startNum,$endNum" else "$endNum,$startNum"
        return autoCheckPointMap[key]
    }

    private fun clearScreenAndDrawList() {
        canvas.drawColor(Color.TRANSPARENT, PorterDuff.Mode.CLEAR)
        for (pair in lineList) {
            canvas.drawLine(
                pair.first.centerX.toFloat(), pair.first.centerY.toFloat(),
                pair.second.centerX.toFloat(), pair.second.centerY.toFloat(), paint
            )
        }
    }

    private fun drawErrorPathTip() {
        canvas.drawColor(Color.TRANSPARENT, PorterDuff.Mode.CLEAR)
        paint.color = Color.rgb(154, 7, 21)
        for (pair in lineList) {
            pair.first.pointState = Constants.POINT_STATE_WRONG
            pair.second.pointState = Constants.POINT_STATE_WRONG
            canvas.drawLine(
                pair.first.centerX.toFloat(), pair.first.centerY.toFloat(),
                pair.second.centerX.toFloat(), pair.second.centerY.toFloat(), paint
            )
        }
        invalidate()
    }

    interface GestureCallBack {
        fun onGestureCodeInput(inputCode: String)
        fun checkedSuccess()
        fun checkedFail()
    }
}
